File: 卡码/代码随想录/1/03-哈希表/Python/07-三数之和/LeetCode_0015.py
#15. 三数之和
from typing import *

class Solution:
	def threeSum(self, nums: List[int]) -> List[List[int]]:
		res=[]
		nums.sort()
		for i in range(len(nums)-2):
			n1=nums[i]
			if n1>0:
				break
			if i>0 and nums[i]==nums[i-1]:
				continue
			left,right=i+1,len(nums)-1
			while left<right:
				n2,n3=nums[left],nums[right]
				if n1+n2+n3==0:
					res.append([n1,n2,n3])
					while left<right and nums[left]==n2:
						left+=1
					while left<right and nums[right]==n3:
						right-=1
				elif n1+n2+n3<0:
					left+=1
				else:
					right-=1
		return res

File: 卡码/代码随想录/1/03-哈希表/Js/07-三数之和/LeetCode_0015.js
//15. 三数之和
const threeSum = function(nums) {
  nums.sort((a,b)=>a-b);
  let res=new Array();
  for(let i=0;i<nums.length-2;i++){
    let n1=nums[i];
    if(n1>0) break;
    if(i>0&&nums[i]==nums[i-1]) continue;
    let left=i+1,right=nums.length-1;
    while(left<right){
      let n2=nums[left],n3=nums[right];
      if(n1+n2+n3===0){
        res.push([n1,n2,n3]);
        while(left<right&&nums[left]==n2) left++;
        while(left<right&&nums[right]==n3) right--;
      }else if(n1+n2+n3<0) left++;
      else right--;
    }
  }
  return res;
};